diff --git a/pypy-5.0.1-src/pypy/module/cpyext/api.py b/pypy-5.0.1-src/pypy/module/cpyext/api.py
index a04753a..dd6bfe4 100644
--- a/pypy-5.0.1-src/pypy/module/cpyext/api.py
+++ b/pypy-5.0.1-src/pypy/module/cpyext/api.py
@@ -59,8 +59,9 @@ include_dirs = [
 class CConfig:
     _compilation_info_ = ExternalCompilationInfo(
         include_dirs=include_dirs,
-        includes=['Python.h', 'stdarg.h', 'structmember.h'],
+        includes=['stdarg.h', 'structmember.h'],
         compile_extra=['-DPy_BUILD_CORE'],
+        pre_include_bits = ['#include "Python.h"'],
         )
 
 class CConfig2:
