diff -up pypy-pypy-2346207d9946/pypy/translator/platform/linux.py.configure-fedora pypy-pypy-2346207d9946/pypy/translator/platform/linux.py
--- pypy-pypy-2346207d9946/pypy/translator/platform/linux.py.configure-fedora	2012-02-09 13:27:19.000000000 -0500
+++ pypy-pypy-2346207d9946/pypy/translator/platform/linux.py	2012-02-10 09:06:20.393066016 -0500
@@ -31,13 +31,14 @@ class BaseLinux(BasePosix):
         return self._pkg_config("libffi", "--libs-only-L",
                                 ['/usr/lib/libffi'])
 
-    def library_dirs_for_libffi_a(self):
-        # places where we need to look for libffi.a
-        # XXX obscuuure!  only look for libffi.a if run with translate.py
-        if 'translate' in sys.modules:
-            return self.library_dirs_for_libffi() + ['/usr/lib']
-        else:
-            return []
+    # Fedora, at least, has the shared version but not the static:
+    #def library_dirs_for_libffi_a(self):
+    #    # places where we need to look for libffi.a
+    #    # XXX obscuuure!  only look for libffi.a if run with translate.py
+    #    if 'translate' in sys.modules:
+    #        return self.library_dirs_for_libffi() + ['/usr/lib']
+    #    else:
+    #        return []
 
 
 class Linux(BaseLinux):
