diff -rup pypy-pypy-f66246c46ca3/rpython/translator/platform/linux.py pypy-pypy-f66246c46ca3/rpython/translator/platform/linux.py
--- pypy-pypy-f66246c46ca3/rpython/translator/platform/linux.py	2013-05-27 10:35:37.679237332 +0200
+++ pypy-pypy-f66246c46ca3/rpython/translator/platform/linux.py	2013-05-27 10:47:41.506354482 +0200
@@ -32,19 +32,20 @@ class BaseLinux(BasePosix):
         return self._pkg_config("libffi", "--libs-only-L",
                                 ['/usr/lib/libffi'])
 
-    def library_dirs_for_libffi_a(self):
-        # places where we need to look for libffi.a
-        # XXX obscuuure!  only look for libffi.a if run with translate.py
-        if 'translate' in sys.modules:
-            if sys.maxint > 2**32:
-                host = 'x86_64'
-            else:
-                host = 'x86'
-            return self.library_dirs_for_libffi() + [
-                '/usr/lib',
-                '/usr/lib/%s-linux-gnu/' % host]
-        else:
-            return []
+    #Fedora, at least, has the shared version but not the static:
+    #def library_dirs_for_libffi_a(self):
+    #    # places where we need to look for libffi.a
+    #    # XXX obscuuure!  only look for libffi.a if run with translate.py
+    #    if 'translate' in sys.modules:
+    #        if sys.maxint > 2**32:
+    #            host = 'x86_64'
+    #        else:
+    #            host = 'x86'
+    #        return self.library_dirs_for_libffi() + [
+    #            '/usr/lib',
+    #            '/usr/lib/%s-linux-gnu/' % host]
+    #    else:
+    #        return []
 
 
 class Linux(BaseLinux):
