diff -rup pypy-pypy-f66246c46ca3/lib-python/2.7/test/test_subprocess.py pypy-pypy-f66246c46ca3/lib-python/2.7/test/test_subprocess.py
--- pypy-pypy-f66246c46ca3/lib-python/2.7/test/test_subprocess.py	2013-05-27 10:35:37.762237806 +0200
+++ pypy-pypy-f66246c46ca3/lib-python/2.7/test/test_subprocess.py	2013-05-28 10:13:03.182536196 +0200
@@ -587,7 +587,7 @@ class ProcessTestCase(BaseTestCase):
         for i in range(1024):
             # Windows raises IOError.  Others raise OSError.
             with self.assertRaises(EnvironmentError) as c:
-                subprocess.Popen(['nonexisting_i_hope'],
+                subprocess.Popen(['/usr/bin/nonexisting_i_hope'],
                                  stdout=subprocess.PIPE,
                                  stderr=subprocess.PIPE)
             # ignore errors that indicate the command was not found
